### ISSUE TITLE
MrCMS.Web Navigation

### ISSUE THREAD

**Issue by zentrendstrading** [MrCMS/MrCMS#54]
I used to be able to cycle through the page.FrontEndRoles in the Navigation.cshtml to hide menu navbar items based on User roles.  Now I can't because FrontEndRoles is no longer defined/available from the page object (NavigationList<NavigationRecord>).  Not sure what I was able to do before.  Any suggestions?

**Comment by zentrendstrading** [comment]
I figured out the hiding of the navigation menu items based on user roles.  Somehow I lost the code to do this in the GetNavigationRecords method.
`public override object GetModel(Navigation widget)
        {
            var rootPages = GetPages(null);
            var childPages = widget.IncludeChildren ? GetPages(rootPages) : new List<Webpage>();
            var navigationRecords =
                rootPages.Where(webpage => webpage.Published).OrderBy(webpage => webpage.DisplayOrder)
                       .Select(webpage => new NavigationRecord
                       {
                           Text = MvcHtmlString.Create(webpage.Name),
                           Url = MvcHtmlString.Create("/" + webpage.LiveUrlSegment),
                           FrontEndRoles = webpage.FrontEndRoles,
                           Children = childPages.Where(webpage1 => webpage1.ParentId == webpage.Id)
                                            .Select(webpage1 =>
                                                    new NavigationRecord
                                                    {
                                                        Text = MvcHtmlString.Create(webpage1.Name),
                                                        Url = MvcHtmlString.Create("/" + webpage1.LiveUrlSegment),
                                                        FrontEndRoles = webpage1.FrontEndRoles
                                                    }).ToList()
                       }).ToList();

            return new NavigationList(navigationRecords.ToList());
        }`

Then modify the Navigation.cshtml page to add an IF to the else branch of the For loop.
`if (page.FrontEndRoles == String.Empty ||
                    (User.Identity.IsAuthenticated && (User.IsInRole("Administrator") || User.IsInRole(page.FrontEndRoles))))
                {
                    <li class="@(activePages.Any(s => s == page.Url.ToString()) ? "active" : "")">
                       <a href="@page.Url">@page.Text</a>
                    </li>   
                }`

Make sure you add FrontEndRoles property to NavigationRecord class.

**Comment by zentrendstrading** [comment]
Update - I still had an issue with menu items appearing with the IF condition in the Navigation.cshtml.  Not sure why the ASP.net Identity User.IsInRole method was not working when it was showing all the roles for a given users even though they did not have permissions to that role.  Here's the modified code to get things to work.  I leveraged theIUserPermissionsService interface to make things work checking to see if the user is allowed to view webpage via the IsCurrentUserAllowed(webpage) method call.
`public class GetNavigationRecords : GetWidgetModelBase<Navigation>
    {
        private readonly ISession _session;
        private readonly IUserUIPermissionsService _userUIPermissionsService;
        

        public GetNavigationRecords(ISession session, IUserUIPermissionsService userUIPermissionsService)
        {
            _session = session;
            _userUIPermissionsService = userUIPermissionsService;
        }

        public override object GetModel(Navigation widget)
        {
            var rootPages = GetPages(null);
            var childPages = widget.IncludeChildren ? GetPages(rootPages) : new List<Webpage>();
            var navigationRecords =
                rootPages.Where(webpage => webpage.Published).OrderBy(webpage => webpage.DisplayOrder)
                       .Select(webpage => new NavigationRecord
                       {
                           Text = MvcHtmlString.Create(webpage.Name),
                           Url = MvcHtmlString.Create("/" + webpage.LiveUrlSegment),
                           FrontEndRoles = webpage.FrontEndRoles,
                           IsUserAllowed = _userUIPermissionsService.IsCurrentUserAllowed(webpage),
                           Children = childPages.Where(webpage1 => webpage1.ParentId == webpage.Id)
                                            .Select(webpage1 =>
                                                    new NavigationRecord
                                                    {
                                                        Text = MvcHtmlString.Create(webpage1.Name),
                                                        Url = MvcHtmlString.Create("/" + webpage1.LiveUrlSegment),
                                                        FrontEndRoles = webpage1.FrontEndRoles,
                                                        IsUserAllowed = _userUIPermissionsService.IsCurrentUserAllowed(webpage1)
                                                    }).ToList()
                       }).ToList();

            return new NavigationList(navigationRecords.ToList());
        }`

Added a new property IsUserAllowed to the NavigationRecord.cs
`public class NavigationRecord
    {
        public NavigationRecord()
        {
            Children = new List<NavigationRecord>();
        }
        public MvcHtmlString Text { get; set; }

        public MvcHtmlString Url { get; set; }

        public string FrontEndRoles { get; set; }

        public bool IsUserAllowed { get; set; }

        public List<NavigationRecord> Children { get; set; }
    }`

And finally cleaned up the IF statement in the Navigation.cshtml
` if (page.IsUserAllowed)
                {
                    <li class="@(activePages.Any(s => s == page.Url.ToString()) ? "active" : "")">
                       <a href="@page.Url">@page.Text</a>
                    </li>   
                }`

**Comment by codeendevor** [comment]
Hello zentrendstrading,

Does this work for you? I had a similar question in issue #32

**Comment by zentrendstrading** [comment]
@codeendever yes, with the method I updated 5 days ago, works like a charm. Have been using this solution for about a week now with no issues. The IUserPermissionsService was key. Simplified the coding as well. Did not have to write as much code.

**Comment by zentrendstrading** [comment]
sorry mistype on @codeendevor

**Comment by codeendevor** [comment]
Cool deal thank you for your efforts

**Comment by zentrendstrading** [comment]
@codeendevor no problem. if you need help or more information, let me know.  fyi, you really dont need the FrontEndRoles property in NavigationRecord. I left it in there for possible future use.

**Comment by codeendevor** [comment]
Cool, yes if you look at what i had you will see it turned into a monster real quick. I had made allot of modifications to the code to account for company information in settings as well as added Structured Data, i have also got the newer version of MrCMS working with the commerce. Just not sure how to share it with everyone.
